File: flow/script/lib/download.ts
/**
 * Video download operations via yt-dlp.
 */

import { BINARIES } from "../config/paths.ts";

const YTDLP = BINARIES.ytDlp;

export interface DownloadOptions {
  url: string;
  outputDir: string;
  playlistItems?: string;
  format?: string;
  mergeFormat?: string;
}

export async function downloadVideo(options: DownloadOptions): Promise<string> {
  const {
    url,
    outputDir,
    playlistItems = "1",
    format = "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best",
    mergeFormat = "mp4",
  } = options;

  const args = [
    "--cookies-from-browser", "chrome",
    "--playlist-items", playlistItems,
    "--format", format,
    "--merge-output-format", mergeFormat,
    "--output", `${outputDir}/%(title)s.%(ext)s`,
    "--no-warnings",
    url,
  ];

  const cmd = new Deno.Command(YTDLP, {
    args,
    stdout: "piped",
    stderr: "piped",
  });

  const { code, stdout, stderr } = await cmd.output();
  const out = new TextDecoder().decode(stdout);
  const err = new TextDecoder().decode(stderr);

  if (code !== 0) {
    throw new Error(`yt-dlp exited ${code}\n${err}`);
  }

  // Extract destination from output
  const match = out.match(/\[download\] Destination: (.+)/m) ||
    err.match(/\[download\] Destination: (.+)/m);
  if (match) return match[1].trim();

  // Fallback: newest file in output dir
  const files: string[] = [];
  for await (const entry of Deno.readDir(outputDir)) {
    if (entry.isFile) files.push(entry.name);
  }
  if (files.length === 0) throw new Error("No file found after download");

  files.sort((a, b) => {
    const sa = Deno.statSync(`${outputDir}/${a}`).mtime ?? new Date(0);
    const sb = Deno.statSync(`${outputDir}/${b}`).mtime ?? new Date(0);
    return sb.getTime() - sa.getTime();
  });

  return `${outputDir}/${files[0]}`;
}

File: flow/script_py/lib/download.py
"""
Video download operations via yt-dlp.
"""

import subprocess
from pathlib import Path
from typing import Optional

from config.paths import BINARIES

YTDLP = BINARIES["yt_dlp"]


def download_video(
    url: str,
    output_dir: str,
    playlist_items: str = "1",
    video_format: str = "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best",
    merge_format: str = "mp4",
) -> str:
    """Download video using yt-dlp."""
    args = [
        YTDLP,
        "--cookies-from-browser", "chrome",
        "--playlist-items", playlist_items,
        "--format", video_format,
        "--merge-output-format", merge_format,
        "--output", f"{output_dir}/%(title)s.%(ext)s",
        "--no-warnings",
        url,
    ]

    result = subprocess.run(args, capture_output=True, text=True, timeout=300)
    if result.returncode != 0:
        raise RuntimeError(f"yt-dlp failed: {result.stderr[:500]}")

    # Extract destination from output
    for line in result.stdout.split("\n"):
        match = __import__("re").match(r"\[download\] Destination: (.+)", line)
        if match:
            return match.group(1).strip()

    # Fallback: newest file in output dir
    out_path = Path(output_dir)
    files = [f for f in out_path.iterdir() if f.is_file()]
    if not files:
        raise FileNotFoundError("No file found after download")

    files.sort(key=lambda f: f.stat().st_mtime, reverse=True)
    return str(files[0])
